### ISSUE TITLE
kube_pod_container_resource_requests does not adequately reflect reality in Kube

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kube-state-metrics#1095]
`kube_pod_container_resource_requests` is kind of useful, but not directly on its own.  It is the most naive view of a pod possible, but misses a bunch of important nuances of the pod resource model (wearing my kube architecture hat right now):

In general, there are 3 major phases to a pod lifetime.

1. Pre-scheduling (no nodeName set) - resources the scheduler is trying to place (roughly the Pending phase)
2. Post-scheduling running phase - resources the pod has to set aside (roughly the Pending->Running phases)
3. Terminal pods - pods that have reached conclusion and will never be run again - the kubelet and scheduler ignore these pods.  Succeeded, Failed, and Evicted pods fall into this category.

Within the running phase there are two sub parts - pod initialization, and container execution.  The calculation for the request of a pod resource is `max(max(init_containers), sum(containers))`.  The scheduler is required to schedule that value.  We also need to leave room to deal with ephemeral containers or sidecars in the future.

There are several common queries I as a kube administrator would want to do that I cannot do today:

1. What does the scheduler think the capacity of the node is (pods that are scheduled to a node and non terminal)
2. How much pending capacity do I have (pods that are not yet scheduled)
3. What is the running capacity of a (namespace / container / set of containers / pod / set of pods)
4. What is the usage profile of a running pod RELATIVE to its phase and requests, taking into account initialization and execution (i.e. are my init requests sized too high)

To correctly form the queries above, the user has to join up to four sources - scheduled, pod phase (which is going to be subtly wrong because phase isn't the true boundary), pod terminal state (which is not exposed as a single metric because of the vagaries of the pod lifecycle), and init containers (which is not available at all).

This is a pretty big capacity planning gap I've noticed as I've spent a fair amount of time tuning infrastructure and workload resource usage.  I would suggest that expose a new metric that uses the rules defined on the pod API (which are not specific to the kubelet or scheduler but apply to any component) to enable those queries.

The metric might be `kube_pod_resource_requests` (to imply this is the correct metric to use for a normal human).  I would probably suggest the following labels (we could debate the actual slicing for maximum flexibility):

* `lifecycle` (it's broader than phase, we'd need to establish in the Kube API some official terms but I can handle that): values `Pending`, `Scheduled`, `Terminated`
* `node` (it should be trivial to query the resources the scheduler is allocating to a node): value none means unscheduled, any other value is a node name
* `name`: pod
* `container`: if empty, should be the sum of the pod resources according to the calculation above.  Should include init containers.
* `container_type`: always set if container is set, should be `init`, `container`, `ephemeral`, or `sidecar` (future).  If not set, the allowed values are empty and `current`.  The current value is always set to the effective max given the pods current phase (max(init, sum(containers)) during pending, based on the current running init container while the pod is initializing, sum of all containers and ephemeral containers during execution, and 0 afterwards)
* `type`: Same values as today.

I think this metric should replace container_resource_requests in general because without the context of lifecycle the metric is not useful.

Example queries:

```
# How much pending CPU is not yet scheduled
sum(kube_pod_resource_requests{lifecycle="Pending",container="",type="cpu"})

# How much CPU is scheduled on a given node right now (what does the scheduler see)
sum(kube_pod_resource_requests{lifecycle="Scheduled",container="",type="cpu",node="foo"})

# How much CPU is scheduled on all nodes
sum(kube_pod_resource_requests{lifecycle="Scheduled",container="",type="cpu",node!=""})

# Show usage vs capacity requested for a given pod over the full lifecycle, showing you exactly where the over or under provisioning in any given phase is
kube_pod_resource_requests{container="current",lifecycle="Scheduled",type="cpu",name="NAME",namespace="NS"} - max by (...) (container_cpu_usage_seconds_total{name="NAME",namespace="NS",container=""})
```

/kind feature

**Comment by lilic** [comment]
I am in two minds of this, while I see a huge value in this metric, we have this note in the readme:
> kube-state-metrics is about generating metrics from Kubernetes API objects without modification. This ensures that features provided by kube-state-metrics have the same grade of stability as the Kubernetes API objects themselves. In turn, this means that kube-state-metrics in certain situations may not show the exact same values as kubectl, as kubectl applies certain heuristics to display comprehensible messages. kube-state-metrics exposes raw data unmodified from the Kubernetes API, this way users have all the data they require and perform heuristics as they see fit.

I feel like this might be applying some of heuristics to metrics mentioned above. Do you disagree? We could get this same value with promql queries correct? Or possibly recording rules.

> lifecycle (it's broader than phase, we'd need to establish in the Kube API some official terms but I can handle that): values Pending, Scheduled, Terminated

That label sounds good to me once added to the api.

cc @tariq1890 @brancz

**Comment by smarterclayton** [comment]
I think this is the fundamental tension.  I would argue the current metric is not useful because you can't use it to answer interesting questions correctly about Kube.  A lot of Kube objects have two parts:

1. explicit values in the API
2. the meaning of those values as they apply to the interaction of values between the API

There is one interpretation that 2 is a heuristic, but I would probably disagree.  The API does define "field A + field B means thing C, which things are part of Kube must respect".  I would probably argue that KSM should be judicious about 2 - but if the KSM metric focuses on 1 but then misleads people about 2 that would probably also be dangerous.

If KSM isn't the right place for it, most kube implementors should filter out this metric (cardinality they don't want) and replace it with the correct metric that watches all pods and then reports it based on the interaction between fields.

**Comment by smarterclayton** [comment]
Lili had a good analogy - asking if this was a computed field that we would report.  We have a *few* computed fields (QoS is the best example) and then decided early in the project we would stop doing that.  So Kube explicitly does not report computed fields, but if they existed, they would mirror roughly what I described above.  It's certainly possible the meaning of the computed field would shift over time, but not without breaking API changes (which we would prevent in kube) or long deprecation periods.  

So basically this boils down to "should Kube computed fields that are not materialized in the API but which multiple components are expected to respect via API documentation" be in KSM?  (i.e. scheduler, kubelet, and quota all follow the rules described here)

**Comment by lilic** [comment]
> I would argue the current metric is not useful because you can't use it to answer interesting questions correctly about Kube.

Any resource or metric can be disabled by passing arguments to kube-state-metrics on runtime, so users who find any metric not useful, can remove it before ingestion. I would not remove the container metric in question, I am sure many people use it and it would break a lot of things for people.

> and then decided early in the project we would stop explicitly calculating those as fields in the API (instead, we document what the computed value is)

Interesting, yes that sounds just like what kube-state-metrics tries to do ;) e.g. expose raw metrics based on kube api and hopefully document how to do queries. Granted we are bad at the docs part of things.

> So Kube explicitly does not report computed fields, but they exist.

They exist in docs you mean?

> should Kube computed fields that are not materialized in the API but which multiple components are expected to respect via API documentation" be in KSM?

Can you point to the kube api docs where pod resource requests are documented, e.g. how kube_pod_container_resource_requests would be calculated. As we would potentially need to make adjustments to this metric based on changes in those docs, this would make it easier to maintain these metrics.


Anyways: I understand your point, I would say its worth considering adding some of those things. 

My main concerns are:
- Who maintains these calculated metrics if we they are not purely reflected from the API?
- Where do we draw the limit, once we allow one or two exceptions?

**Comment by smarterclayton** [comment]
> docs as API

https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#resources

> core design as API

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/container-init.md#resources

> API field:

 	// The resourceRequirements of an init container are taken into account during scheduling
	// by finding the highest request/limit for each resource type, and then using the max of
	// of that value or the sum of the normal containers. Limits are applied to init containers
	// in a similar fashion.

I think we could change containers to describe the resource behavior (https://github.com/kubernetes/kubernetes/pull/89115).

> Who maintains these calculated metrics if we they are not purely reflected from the API?

Maintainers of the subsystems who expect prometheus metrics to reflect the Kube system model?  There definitely has to be ownership - should a Kube maintainer be allowed to suggest these metrics are important if they are wiling to maintain them?

> Where do we draw the limit, once we allow one or two exceptions?

One option is components that are part of the conformance model for Kubernetes - for instance, to pass conformance you have to schedule and execute pods with the requested resources and obey these rules.  Agree it's not as hard and fast as the rule about "what's in the API"

**Comment by smarterclayton** [comment]
Prototyping the rough structure https://github.com/openshift/kube-state-metrics/pull/27 - will provide prometheus dumps with examples so folks can see what I mean.

**Comment by smarterclayton** [comment]
Something that I realized with testing and discussed with Lili - KSM is often tuned to be slow scrape because of volume of data.  That's another dimension where this might belong more in the scheduler or KCM, in which case the concerns about maintenance are reduced, although cardinality is high.  I'm going to talk to sig scheduling about this and then come back to sig-instrumentation with some questions / options.

**Comment by brancz** [comment]
Note high scrape intervals is a thing from the past before we did the performance improvements of ksm, where a scrape could take up to 30s, now even in huge clusters we're under 15s, so it's pretty safe to go back to 15s scrape intervals. Remember that with cardinality the expensive thing is tracking the individual metrics themselves not so much the amount of samples, as they compress better with more samples.

**Comment by brancz** [comment]
To be clear I do think all of this can already be done with recording rules and with a lower scrape interval. I do agree high level recording rules for better experience are a good idea. But experience has shown that any pre computation we do tends to have bugs and then invalidates historic data which is why I prefer to stay with the “rawest” data we can get.

I totally agree that there is a bit of excessive exposition of metrics on kube state metrics (some justified some less). The point is users can whitelist only those metrics they want and get their own best value out of it that way.

**Comment by smarterclayton** [comment]
I don't think this can or should be done with recording rules.  This is data derived from the core algorithms of the Kube system and it's much more nuanced than even the stuff in KSM gets close to - in this case the rawest data is stuff derived in calculations from internal libraries.  I don't think its appropriate to reproduce internal calculations in recording rules - like apiserver latencies, cgroup to pod latencies, etc.  Here's the actual calculation that roughly mirrors the quota and scheduler subsystem calculations (as the pod lifecycle defines them): https://github.com/openshift/origin/pull/24713/files#diff-5872fa366ad2573f88d5eed74fc3ef23R155.  I don't think it would be possible to do the correct calculation for the running init container from a metric.

Here's some example runs from CI https://storage.googleapis.com/origin-ci-test/pr-logs/pull/24713/pull-ci-openshift-origin-master-e2e-gcp/6422/artifacts/e2e-gcp/metrics/prometheus.tar

2020-03-18 20:37

Memory requests by node as the scheduler sees it for running workloads

```
sum by (node,resource) (max(label_replace(label_replace(scheduler_pod_resources{lifecycle=~"Initializing|Running",resource="memory",type="requests"}, "namespace","$1","exported_namespace","(.*)"), "name", "$1", "exported_pod", "(.*)")) without (instance,endpoint,job,exported_namespace,exported_pod,pod,service))
```

Memory requests by namespace, without pending resources

```
sum by (namespace,resource) (max(label_replace(label_replace(scheduler_pod_resources{lifecycle=~"Initializing|Running",resource="memory",type="requests"}, "namespace","$1","exported_namespace","(.*)"), "name", "$1", "exported_pod", "(.*)")) without (instance,endpoint,job,exported_namespace,exported_pod,pod,service))
```

Memory requests by namespace including pending resources

```
sum by (namespace,resource) (max(label_replace(label_replace(scheduler_pod_resources{lifecycle="",resource="memory",type="requests"}, "namespace","$1","exported_namespace","(.*)"), "name", "$1", "exported_pod", "(.*)")) without (instance,endpoint,job,exported_namespace,exported_pod,pod,service))
```

Unscheduled pod memory requests by namespace

```
sum by (namespace,resource) (max(label_replace(label_replace(scheduler_pod_resources{lifecycle="Pending",resource="memory",type="requests"}, "namespace","$1","exported_namespace","(.*)"), "name", "$1", "exported_pod", "(.*)")) without (instance,endpoint,job,exported_namespace,exported_pod,pod,service))
```

CPU cores requested by initializing containers vs running containers

```
sum by (namespace,resource) (max(label_replace(label_replace(scheduler_pod_resources{lifecycle="Initializing",resource="cpu",type="requests"}, "namespace","$1","exported_namespace","(.*)"), "name", "$1", "exported_pod", "(.*)")) without (instance,endpoint,job,exported_namespace,exported_pod,pod,service))
```

(unfortunately the workload from e2e is neither long running enough nor complicated enough to show the places where initialized requests may be lower or higher than running requests, like they would in production systems using init containers)

Pod requests - usage as nodes and scheduler see it (this the aggregate over requested memory for pods). If we had allocatable as a metric (which is really the critical piece) we would have slack scheduling capacity.

```
sum by (node) (max(label_replace(label_replace(scheduler_pod_resources{lifecycle=~"Initializing|Running",resource="memory",type="requests"}, "namespace","$1","exported_namespace","(.*)"), "name", "$1", "exported_pod", "(.*)")) without (instance,endpoint,job,exported_namespace,exported_pod,pod,service)) - max by (node) (container_memory_rss{id="/kubepods.slice"})
```

**Comment by smarterclayton** [comment]
Frankly, I'm kind of embarrassed Kube hasn't reported these metrics before.  Most people are relying on wrong or incorrect calculations based on part of this model, so they're making subtly (or not so subtly) flawed judgements.

**Comment by lilic** [comment]
> I don't think this can or should be done with recording rules.

Basically I gather, from this and our other conversations, what you want is the metric to be in kube scheduler itself. I agree kube-state-metrics would not provide precise enough metrics, neither would recording rules. So this belongs in core kube scheduler code as you demonstrated in the origin PR. Best to bring this up in SIG instrumentation, don't think you would have any problems convincing people this is needed or much objections.

Think we can close this issue as we all agree it belongs in the scheduler, unless there is any objections? :) 

Side note about the metric itself:
If our goal is to reflect the scheduling decisions the container level details don't make sense. (kube_scheduler_pod_resources metric has container label)

**Comment by brancz** [comment]
I agree that if our goal is to reflect _scheduling state_ then the _scheduler_ would be the right place :slightly_smiling_face: .

**Comment by smarterclayton** [comment]
The biggest question is then what to do about alternative schedulers.  I'll draft a discussion for both sig-scheduling and sig-instrumentation and we can take it to a meeting.

**Comment by brancz** [comment]
Yeah that makes things more complicated indeed.

**Comment by lilic** [comment]
As there is a KEP for this in Kubernetes, I believe we can just close this? Or do we want to leave it open and remove this metric once it lands in Kubernetes?

**Comment by smarterclayton** [comment]
Note that this is beta in kube 1.21 as `kube_pod_resource_request` and `kube_pod_resource_limit` - feedback especially welcome to set decisions on GA timeline.
